libraries/chain/wasm_interface_collection.cpp

```diff
@@ -0,0 +1,131 @@
+   thread_local static eosvmoc::memory mem;
+};
+
+thread_local std::unique_ptr<eosvmoc::executor> eosvmoc_tier::exec{};
+thread_local eosvmoc::memory eosvmoc_tier::mem{wasm_constraints::maximum_linear_memory / wasm_constraints::wasm_page_size};
+#endif
+
+wasm_interface_collection::wasm_interface_collection(wasm_interface::vm_type vm, wasm_interface::vm_oc_enable eosvmoc_tierup,
+                                                     const chainbase::database& d, const std::filesystem::path& data_dir,
+                                                     const eosvmoc::config& eosvmoc_config, bool profile)
+   : main_thread_id(std::this_thread::get_id())
+   , wasm_runtime(vm)
+   , eosvmoc_tierup(eosvmoc_tierup)
+   , wasmif(vm, d, data_dir, eosvmoc_config, profile) {
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+   if (eosvmoc_tierup != wasm_interface::vm_oc_enable::oc_none) {
+      EOS_ASSERT(vm != wasm_interface::vm_type::eos_vm_oc, wasm_exception, "You can't use EOS VM OC as the base runtime when tier up is activated");
+      eosvmoc = std::make_unique<eosvmoc_tier>(data_dir, eosvmoc_config, d);
+   }
+#endif
+}
+
+wasm_interface_collection::~wasm_interface_collection() = default;
+
+void wasm_interface_collection::apply(const digest_type& code_hash, const uint8_t& vm_type, const uint8_t& vm_version, apply_context& context) {
+   if (substitute_apply && substitute_apply(code_hash, vm_type, vm_version, context))
+      return;
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+   if (eosvmoc && (eosvmoc_tierup == wasm_interface::vm_oc_enable::oc_all || context.should_use_eos_vm_oc())) {
+      const chain::eosvmoc::code_descriptor* cd = nullptr;
+      chain::eosvmoc::code_cache_base::get_cd_failure failure = chain::eosvmoc::code_cache_base::get_cd_failure::temporary;
+      try {
+         const bool high_priority = context.get_receiver().prefix() == chain::config::system_account_name;
+         cd = eosvmoc->cc.get_descriptor_for_code(high_priority, code_hash, vm_version, context.control.is_write_window(), failure);
+         if (test_disable_tierup)
+            cd = nullptr;
+      } catch (...) {
+         // swallow errors here, if EOS VM OC has gone in to the weeds we shouldn't bail: continue to try and run baseline
+         // In the future, consider moving bits of EOS VM that can fire exceptions and such out of this call path
+         static bool once_is_enough;
+         if (!once_is_enough)
+            elog("EOS VM OC has encountered an unexpected failure");
+         once_is_enough = true;
+      }
+      if (cd) {
+         if (!context.is_applying_block()) // read_only_trx_test.py looks for this log statement
+            tlog("${a} speculatively executing ${h} with eos vm oc", ("a", context.get_receiver())("h", code_hash));
+         eosvmoc->exec->execute(*cd, eosvmoc->mem, context);
+         return;
+      }
+   }
+#endif
+   if (is_on_main_thread()) {
+      wasmif.apply(code_hash, vm_type, vm_version, context);
+      return;
+   }
```

Comment:
why not an else here?

Reply:
The rest of the method short-cuts out with `return`. So this follows the rest of the method approach.